### PR TITLE
Implement meta/hash and meta/hash256 endpoints.

### PR DESCRIPTION
**Pull request by frankban**
Those are used to retrieve archive hashes for entities.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/759/
Test PASSed.

File: internal/legacy/api_test.go [hunk 1]
```diff
@@ -329,61 +330,27 @@ func (s *APISuite) TestCharmPackageCharmInfo(c *gc.C) {
...
-		ExpectStatus: http.StatusOK,
-		ExpectBody: map[string]charm.InfoResponse{
-			wordpressURL.String(): {
-				CanonicalURL: wordpressURL.String(),
-				Sha256:       sum256,
-				Revision:     0,
+	// TODO frankban: remove this test after updating entities in the
+	// production db with their SHA256 hash value. Entities are updated by
+	// running the cshash256 command.
+	id, ch := s.addCharm(c, "wordpress", "cs:precise/wordpress-0")
```
Comment (mhilton) [inline internal/legacy/api_test.go]: Can you make this test use a charm with a user please? that way it won't break when the promulgation stuff merges.

File: internal/v4/api_test.go [hunk 2]
```diff
@@ -1749,6 +1772,29 @@ func (s *APISuite) TestDebugPprofFailsWithoutAuth(c *gc.C) {
 	}
 }
 
+func (s *APISuite) TestHash256Laziness(c *gc.C) {
+	// TODO frankban: remove this test after updating entities in the
+	// production db with their SHA256 hash value. Entities are updated by
+	// running the cshash256 command.
+	id, _ := s.addCharm(c, "wordpress", "cs:precise/wordpress-0")
```
Comment (mhilton) [inline internal/v4/api_test.go]: And add a user to this one too please.

**Comment by mhilton** [comment]
:+1: LGTM with a couple of small changes.

File: internal/charmstore/hash.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	// or fails, or if several instances of the charm store do it concurrently,
+	// and it doesn't need to be on the critical path for API endpoints.
+	go UpdateEntitySHA256(s, id, sum256)
+
+	return sum256, nil
+}
+
+// UpdateEntitySHA256 updates the BlobHash256 entry for the entity.
+// It is defined as a variable so that it can be mocked in tests.
+// This function will be removed soon.
```
Comment (bac) [inline internal/charmstore/hash.go:44]: Great explanation, thank you.

**Comment by bac** [comment]
:+1:

File: internal/legacy/api_test.go (hunk 1 above)
Comment (frankban) [inline internal/legacy/api_test.go]: Done.

File: internal/v4/api_test.go (hunk 2 above)
Comment (frankban) [inline internal/v4/api_test.go]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/760/
Test PASSed.

**Comment by frankban** [comment]
Thanks for the reviews Martin and Brad!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**